test/e2e/common.go

```diff
@@ -338,24 +330,119 @@ func HasRootOnDisk(output string) bool {
 	cmd := "df -h"
 	output, err := executeSSHCommand(client, cmd)
 	if err != nil {
 		return false, fmt.Errorf("error executing 'df -h': %w", err)
 	}
 
-	isDisk := HasRootOnDisk(output)
-	if isDisk {
+	bootedFromDisk = HasRootOnDisk(output)
+	if bootedFromDisk {
 		capm3_e2e.Logf("System is booted from a disk.")
 	} else {
 		capm3_e2e.Logf("System is booted from a live ISO.")
 	}
 
-	return isDisk, nil
+	return bootedFromDisk, nil
+}
+
+func EstablishSSHConnection(authMethod, user, address string, useKeyAuth bool) *ssh.Client {
+	var auth ssh.AuthMethod
+	if useKeyAuth {
+		key, err := os.ReadFile(authMethod) // #nosec G304
+		Expect(err).NotTo(HaveOccurred(), "unable to read private key")
+
+		signer, err := ssh.ParsePrivateKey(key)
+		Expect(err).NotTo(HaveOccurred(), "unable to parse private key")
+		auth = ssh.PublicKeys(signer)
+	} else {
+		auth = ssh.Password(authMethod)
+	}
+
+	config := &ssh.ClientConfig{
+		User:            user,
+		Auth:            []ssh.AuthMethod{auth},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // #nosec G106
+	}
+
+	var client *ssh.Client
+	var err error
+	Eventually(func() error {
+		client, err = ssh.Dial("tcp", address, config)
+		return err
+	}, 2*time.Minute, 10*time.Second).Should(Succeed(), "Failed to establish SSH connection")
+
+	return client
+}
+
+// CreateUserdataSecret creates a Kubernetes secret intended for cloud-init usage.
+// This userdata is utilized during BMH's initialization and setup.
+func CreateUserdataSecret(ctx context.Context, client client.Client, namespace string, secretName string, sshPubKeyPath string) {
+	sshPubKeyData, err := os.ReadFile(sshPubKeyPath) // #nosec G304
+	Expect(err).NotTo(HaveOccurred(), "Failed to read SSH public key file")
+
+	userDataContent := fmt.Sprintf(`#!/bin/sh
+mkdir /root/.ssh
+mkdir /home/cirros/.ssh
+chmod 700 /root/.ssh
+chmod 700 /home/cirros/.ssh
+chown cirros /home/cirros/.ssh
+echo "%s" >> /home/cirros/.ssh/authorized_keys
+echo "%s" >> /root/.ssh/authorized_keys`, sshPubKeyData, sshPubKeyData)
+
+	CreateSecret(ctx, client, namespace, secretName, map[string]string{"userData": userDataContent})
+}
+
+// CreateMetadataSecret creates a Kubernetes secret intended for cloud-init usage.
+// This metadata is utilized during BMH's initialization and setup.
+func CreateMetadataSecret(ctx context.Context, client client.Client, namespace string, secretName string) {
+	metaDataContent := `{
+		"instance-id": "cirros-bmo-e2e-test",
+		"local-hostname": "cirros-bmo-test-vm"
+	}`
+
+	CreateSecret(ctx, client, namespace, secretName, map[string]string{"metaData": metaDataContent})
+}
+
+// PerformSSHBootCheck performs an SSH check to verify the node's boot source.
+// The `expectedBootMode` parameter should be "disk" or "memory".
+// The `useKeyAuth` flag specifies if SSH key authentication should be used.
+func PerformSSHBootCheck(e2eConfig *Config, expectedBootMode string, useKeyAuth bool) {
```

Comment:
I think it would be nicer to have a parameter for the `ssh.AuthMethod` here also instead of the `useKeyAuth`.